pallets/settlement/src/lib.rs

```diff
@@ -1522,55 +1524,101 @@ impl<T: Trait> Module<T> {
+    pub fn affirm_and_may_schedule_instruction(
+        origin: <T as frame_system::Trait>::Origin,
+        instruction_id: u64,
+        portfolios: Vec<PortfolioId>,
+    ) -> DispatchResult {
+        Self::base_affirm_instruction(origin, instruction_id, portfolios)?;
+        // Schedule the instruction if conditions are met
+        Self::maybe_schedule_instruction(
+            Self::instruction_affirms_pending(instruction_id),
             Self::instruction_details(instruction_id).settlement_type,
             instruction_id,
-        );
+        )
+    }
+
+    /// It is a specialized function to affirm & execute the instruction when all affirmations
+    /// have received.
+    ///
+    /// NB - It is recommended to use this function only in the STO pallet to support the DVP settlement.
+    pub fn affirm_and_execute_instruction(
+        origin: <T as frame_system::Trait>::Origin,
+        instruction_id: u64,
+        portfolios: Vec<PortfolioId>,
+    ) -> DispatchResult {
+        Self::base_affirm_instruction(origin, instruction_id, portfolios)?;
+        Self::is_execute_instruction(
+            instruction_id,
+            Self::instruction_affirms_pending(instruction_id),
+            Self::instruction_details(instruction_id).settlement_type,
+        )
+    }
+
+    /// It is a specialized function to affirm with receipts & execute the instruction when all affirmations
+    /// have received.
+    ///
+    /// NB - It is recommended to use this function only in the STO pallet to support the DVP settlement.
+    pub fn affirm_with_receipts_and_execute_instruction(
+        origin: <T as frame_system::Trait>::Origin,
+        instruction_id: u64,
+        receipt_details: Vec<ReceiptDetails<T::AccountId, T::OffChainSignature>>,
+        portfolios: Vec<PortfolioId>,
+    ) -> DispatchResult {
+        Self::base_affirm_with_receipts(origin, instruction_id, receipt_details, portfolios)?;
+        Self::is_execute_instruction(
+            instruction_id,
+            Self::instruction_affirms_pending(instruction_id),
+            Self::instruction_details(instruction_id).settlement_type,
+        )
+    }
 
-        weight_for_instruction_execution.map(|info| {
-            info.actual_weight
-                .map(|w| w.saturating_add(weight_for::weight_for_affirmation_instruction::<T>()))
-                .into()
-        })
+    fn is_execute_instruction(
+        instruction_id: u64,
+        affirms_pending: u64,
+        settlement_type: SettlementType<T::BlockNumber>,
+    ) -> DispatchResult {
+        // As this function will assumes that the settlement type will always be `SettleOnAffirmation` but to be
+        // defensive it is good to check before executing the instruction.
+        if settlement_type == SettlementType::SettleOnAffirmation && affirms_pending == 0 {
+            let _ = Self::execute_instruction(instruction_id);
+        }
+        Ok(())
```

Comment:
We should be returning errors if the execution fails or if the settlement type is different so that the STO module can handle those.